lightning/src/ln/channel.rs

```diff
@@ -3395,6 +3784,53 @@ pub(crate) fn per_outbound_htlc_counterparty_commit_tx_fee_msat(feerate_per_kw:
 	}
 }
 
+#[cfg(any(dual_funding, splicing))]
+pub(super) fn maybe_add_funding_change_output<SP: Deref>(
+	signer_provider: &SP, is_initiator: bool, our_funding_satoshis: u64,
+	funding_inputs_prev_outputs: &Vec<TxOut>, funding_outputs: &mut Vec<OutputOwned>,
+	funding_feerate_sat_per_1000_weight: u32, total_input_satoshis: u64,
+	holder_dust_limit_satoshis: u64, channel_keys_id: [u8; 32],
+) -> Result<bool, ChannelError> where
+	SP::Target: SignerProvider,
+{
+	let our_funding_inputs_weight = funding_inputs_prev_outputs.iter().fold(0u64, |weight, prev_output| {
+		weight.saturating_add(estimate_input_weight(prev_output).to_wu())
+	});
+	let our_funding_outputs_weight = funding_outputs.iter().fold(0u64, |weight, output| {
+		weight.saturating_add(get_output_weight(&output.tx_out().script_pubkey).to_wu())
+	});
+	let our_contributed_weight = our_funding_outputs_weight.saturating_add(our_funding_inputs_weight);
+	let mut fees_sats = fee_for_weight(funding_feerate_sat_per_1000_weight, our_contributed_weight);
+
+	// If we are the initiator, we must pay for weight of all common fields in the funding transaction.
+	if is_initiator {
+		let common_fees = fee_for_weight(funding_feerate_sat_per_1000_weight, TX_COMMON_FIELDS_WEIGHT);
+		fees_sats = fees_sats.saturating_add(common_fees);
+	}
+
+	let remaining_value = total_input_satoshis
+		.saturating_sub(our_funding_satoshis)
+		.saturating_sub(fees_sats);
+
+	if remaining_value < holder_dust_limit_satoshis {
+		Ok(false)
+	} else {
+		let change_script = signer_provider.get_destination_script(channel_keys_id).map_err(
```

Comment:
If I understand correctly we want the user to pass in a change script instead of doing this? I was under the impression we'd have different addresses on each call of `get_destination_script()` as specified in the API.

Reply:
Right so it's a pure function, so we'd have address re-use :(
Ok we'll need to pass through the funding context.

Reply:
And if we didn't bother with having an explicit "funding satoshis" in the API then we wouldn't need to worry about the change script.